hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -104,76 +103,82 @@
+            injectDependencies(defaultExpiryPolicy);
+        } else {
+            throw new IllegalStateException("Expiry policy factory cannot be null!");
         }
-        final Factory<ExpiryPolicy> expiryPolicyFactory = cacheConfig.getExpiryPolicyFactory();
-        this.defaultExpiryPolicy = expiryPolicyFactory.create();
 
-        this.records = createRecordCacheMap();
-        this.maxSizeChecker = createCacheMaxSizeChecker(evictionConfig.getSize(), evictionConfig.getMaximumSizePolicy());
-        this.evictionPolicyEvaluator = createEvictionPolicyEvaluator(evictionConfig);
-        this.evictionChecker = createEvictionChecker(evictionConfig);
-        this.evictionStrategy = createEvictionStrategy(evictionConfig);
+        cacheContext = cacheService.getOrCreateCacheContext(name);
+        records = createRecordCacheMap();
+        maxSizeChecker = createCacheMaxSizeChecker(evictionConfig.getSize(), evictionConfig.getMaximumSizePolicy());
+        evictionPolicyEvaluator = createEvictionPolicyEvaluator(evictionConfig);
+        evictionChecker = createEvictionChecker(evictionConfig);
+        evictionStrategy = createEvictionStrategy(evictionConfig);
 
-        // Register "cacheWriter" if it is "Closable" to be closed while cache is being destroyed
-        if (cacheWriter instanceof Closeable) {
-            cacheService.addCacheResource(name, (Closeable) cacheWriter);
-        }
-        // Register "cacheLoader" if it is "Closable" to be closed while cache is being destroyed
-        if (cacheLoader instanceof Closeable) {
-            cacheService.addCacheResource(name, (Closeable) cacheLoader);
-        }
-        // Register "defaultExpiryPolicy" if it is "Closable" to be closed while cache is being destroyed
-        if (defaultExpiryPolicy instanceof Closeable) {
-            cacheService.addCacheResource(name, (Closeable) defaultExpiryPolicy);
-        }
+        registerResourceIfItIsClosable(cacheWriter);
+        registerResourceIfItIsClosable(cacheLoader);
+        registerResourceIfItIsClosable(defaultExpiryPolicy);
 
         init();
     }
     //CHECKSTYLE:ON
 
     private boolean isPrimary() {
-        IPartitionService partitionService = nodeEngine.getPartitionService();
-        IPartition partition = partitionService.getPartition(partitionId, false);
+        IPartition partition = nodeEngine.getPartitionService().getPartition(partitionId, false);
         Address owner = partition.getOwnerOrNull();
         Address thisAddress = nodeEngine.getThisAddress();
         return owner != null && owner.equals(thisAddress);
     }
 
+    private void injectDependencies(Object obj) {
+        if (obj instanceof HazelcastInstanceAware) {
+            ((HazelcastInstanceAware) obj).setHazelcastInstance(nodeEngine.getHazelcastInstance());
+        }
+    }
+
+    private void registerResourceIfItIsClosable(Object resource) {
+        if (resource instanceof Closeable) {
```

Comment:
being a `Closable` is not mandatory in here. Therefore, there should not be an exception if it is not a `Closable`